### ISSUE TITLE
Language Server 2.0 - Use y.js for communition

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#5419]
## TL;DR;

- this issue came up with many ideas, but
- #14142 is the **"winner"**
- once #14438 is integrated (it is green as of Mar 1, 2026)
- all the communication will be handled via Y.js server

## Language Server 1.0
The current version of the language server (LS) does not have an optimal performance. Back in the days it was built with the different requirements in mind, and can be implemented in a more optimal way.

## Language Server 2.0
A new language server can be created with different requirements based on the experience of building the first version.

- Focus on performance. Performance should be the main requirement when designing a new LS. 
- Switch to binary protocol. The text protocol is a legacy, and nothing prevents us to switching to more efficient binary format.
- Drop multi-user support. LS 1.0 can support multiple users but we've never used it this way. Removing this requirement should drastically simplify the logic. Experiments with [y.js sharing has been made](https://github.com/orgs/enso-org/discussions/7722#discussioncomment-6905426)
- Switch to in-memory protocol between the LS and the Engine. In LS 1.0 we have to serialize-deserialize message every time it crosses the LS/Engine boundary. In-memory protocol will eliminate the overhead of serialization.

## Use Virtual Threads

JDK21 is the first LTS version to provide support for [virtual threads](https://openjdk.org/jeps/425). Virtual threads are light to create and light to block. One can write sequential code and invoke it as effectively as reactive code. We should use one of the frameworks that provides support for virtual threads and simplify the current actor based system. Consider: 
- Helidon Níma: https://www.youtube.com/watch?v=aYY04D4_OQA
- Micronaut: https://github.com/micronaut-projects/micronaut-core/pull/8180
- Quarkus: https://quarkus.io/guides/virtual-threads

Micronaut & Quarkus rely a lot on dependency injection. It may be impossible to use them without all the DI magic. Níma provides [imperative API to build and handle the routes](https://github.com/helidon-io/helidon/blob/4.0.0-M1/examples/nima/quickstart/src/main/java/io/helidon/examples/nima/quickstart/QuickstartMain.java#L60).

## Distribute Language Server with the IDE

There is an [effort to run language server as hosted Java](https://github.com/enso-org/enso/issues/7683#issuecomment-1698818418). It'd be great if these two efforts emerged together.

## Write Language Server in JavaScript or Enso

An alternative to above mentioned Java frameworks is to write language server in JavaScript and execute it by GraalVM. We cannot easily support all `node.js` modules, but we can support latest, greatest JavaScript specification and emulate essential `node.js` APIs (for example for handling [websockets](https://github.com/theturtle32/WebSocket-Node/blob/cce6d468986dd356a52af5630fd8ed5726ba5b7a/docs/WebSocketServer.md) via `WebSocketServer`). Enso Engine is already running JavaScript (to process JSON), so running JavaScript for language server isn't going to add much overhead.

Alternatively the language server could be written in Enso itself. Using the same emulation of _websockets_ we could handle the incoming and outgoing messages in plan Enso code. Using JavaScript or Enso comes with familiarity of the languages and also support the _Distribute Language Server with the IDE_ goal - having it in JS or Enso will require less changes in IDE build system than writing the server in Java.

**Comment by hubertp** [comment]
Another thing to consider is that we cannot upgrade to latest Akka nor should we base any further development on it because of its restrictive license change - https://www.lightbend.com/akka/license.

<blockquote><img src="https://downloads.lightbend.com/website/social/lightbend-logo-1024-576.jpg" width="48" align="right"><div>Lightbend</div><div><strong><a href="https://www.lightbend.com/akka/license">Akka BSL 1.1 License | @lightbend</a></strong></div></blockquote>

**Comment by wdanilo** [comment]
> Another thing to consider is that we cannot upgrade to latest Akka nor should we base any further development on it because of its restrictive license change - https://www.lightbend.com/akka/license.
> 
> > <img alt="" width="48" src="https://camo.githubusercontent.com/636694e824cf84872fd77da79d6d5849a2310fcd625cc6bea68d17b810947990/68747470733a2f2f646f776e6c6f6164732e6c6967687462656e642e636f6d2f776562736974652f736f6369616c2f6c6967687462656e642d6c6f676f2d313032342d3537362e6a7067">Lightbend**[Akka BSL 1.1 License | @lightbend](https://www.lightbend.com/akka/license)**

Oh wow. I did not dig deep enough to understand the new license, but are you sure that such projects as our would not be able to use it? Looks like a killing license for Akka.

<blockquote><img src="https://downloads.lightbend.com/website/social/lightbend-logo-1024-576.jpg" width="48" align="right"><div>Lightbend</div><div><strong><a href="https://www.lightbend.com/akka/license">Akka BSL 1.1 License | @lightbend</a></strong></div></blockquote>

**Comment by hubertp** [comment]
@wdanilo Yeah, fairly sure. Here is more on the topic https://www.infoq.com/news/2022/09/akka-no-longer-open-source/. There is a whole discussion on forums about it when it happened but the general sentiment is toward moving away from it.

We are OK for now, but in the longer term it won't make sense to base more infrastructure on it.
@JaroslavTulach was suggesting Micronaut which would be inline with reducing Scala presence in the codebase.

<blockquote><img src="https://res.infoq.com/news/2022/09/akka-no-longer-open-source/en/headerimage/generatedHeaderImage-1663609531852.jpg" width="48" align="right"><div>InfoQ</div><div><strong><a href="https://www.infoq.com/news/2022/09/akka-no-longer-open-source/">Lightbend Changes Akka License and Is No Longer Open Source </a></strong></div><div>Akka, a toolkit for writing concurrent distributed applications based on the actor model, was created thirteen years ago by Jonas Bonér, founder and CEO at Lightbend. The company has recently announced a new Akka license model that has changed from the open-source  Apache 2.0 to the source-available Business Source License (BSL) 1.1.</div></blockquote>

**Comment by JaroslavTulach** [comment]
> @wdanilo wrote:
> Oh wow. I did not dig deep enough to understand the new license, but are you sure that such projects as our would not be able to use it? 

A quote from https://mariadb.com/bsl-faq-adopting/

>    Q: Can I use BSL software with an Open Source project?
>
>    A:  No.

> @wdanilo wrote:
> Looks like a killing license for Akka.

Right. Nobody will ever write any new open source library against Akka, if it is licensed under BSL. Using Akka in an open source project will also become quite painful. What a path into oblivion!

<blockquote><img src="https://mariadb.com/wp-content/uploads/2018/10/mariadb-social-share.jpg" width="48" align="right"><div>MariaDB</div><div><strong><a href="https://mariadb.com/bsl-faq-adopting/">Adopting and Developing BSL Software | MariaDB</a></strong></div></blockquote>

**Comment by wdanilo** [comment]
This is crazy!

**Comment by JaroslavTulach** [comment]
Idea: Use [y.js AST for communition](https://github.com/enso-org/enso/issues/9257#issuecomment-1978474823) between the engine and IDEs.

**Comment by JaroslavTulach** [comment]
- closing in favor or #14142 
- with the hope #14438 will be integrated soon
